### PR TITLE
gc epoch: refactor epoch

### PR DESCRIPTION
**Pull request by runshenzhu**
This PR refactors the epoch code:
- `reclaim_tail_` --> `gc_tail_`, `queue_tail_` --> `readonly_tail_`
- put readonly txn and normal txn into the same epoch (current epoch), but store max_cid seperately. For readonly txn, update `epoch.max_readonly_cid_`

ping @jessesleeping @ranxian 
Signed-off-by: runshenzhu runshen.zhu@gmail.com

### CODE REVIEW

**Comment by jamesrxian** [comment]
Good job

File: src/backend/concurrency/epoch_manager.h [hunk 1]
```diff
@@ -89,15 +91,16 @@ class EpochManager {
...
+    auto epoch = current_epoch_.load();
 
     size_t epoch_idx = epoch % epoch_queue_size_;
     epoch_queue_[epoch_idx].ro_txn_ref_count_++;
 
     // Set the max cid in the tuple
-    auto max_cid_ptr = &(epoch_queue_[epoch_idx].max_cid_);
-    AtomicMax(max_cid_ptr, begin_cid);
-
+    // auto max_cid_ptr = &(epoch_queue_[epoch_idx].max_cid_);
```
Comment (ranxian) [inline src/backend/concurrency/epoch_manager.h]: asd asd

File: src/backend/concurrency/epoch_manager.h [hunk 2]
```diff
@@ -206,29 +209,28 @@ class EpochManager {
       }
 
       // save max cid
-      auto max = epoch_queue_[idx].max_cid_;
-      AtomicMax(&max_cid_gc_, max);
+      auto max_ = std::max(epoch_queue_[idx].max_cid_, epoch_queue_[idx].max_readonly_cid_);
```
Comment (jessesleeping) [inline src/backend/concurrency/epoch_manager.h:212]: I think GC tail should not consider the max_cid_ of the epoch. It only need to look at the max_read_only_cid_ and get the max.

File: src/backend/concurrency/epoch_manager.h [hunk 3]
```diff
@@ -237,23 +239,22 @@ class EpochManager {
 
       auto idx = tail % epoch_queue_size_;
 
-      // inc tail until we find an epoch that has running txn
-      if(epoch_queue_[idx].rw_txn_ref_count_ > 0) {
+      // inc tail until we find an epoch that has running read only txn
+      if(epoch_queue_[idx].ro_txn_ref_count_ > 0) {
```
Comment (jessesleeping) [inline src/backend/concurrency/epoch_manager.h]: For queue tail, this should be rw_txn_ref_count

File: src/backend/concurrency/epoch_manager.h (hunk 2 above)
Comment (jessesleeping) [inline src/backend/concurrency/epoch_manager.h:212]: It's right

**Comment by runshenzhu** [comment]
updated. But I think we should hold this PR, until its correctness is ensured.

**Comment by jessesleeping** [comment]
For experiment purpose.. let's make things simple

**Pull request closed without merging**